src/lib.rs

```diff
@@ -221,6 +235,51 @@ macro_rules! get_function {
 get_function!(name: get_mut, mutability: mut);
 get_function!(name: get, mutability: );
 
+/// Macro to parametrise over mutability in get_node methods.
+macro_rules! get_node_function {
+    (
+        name: $name:ident,
+        mutability: $($mut_:tt)*
+    ) => {
+        id!(fn $name<'a, K, V>(
+            trie: &'a $($mut_)* Trie<K, V>,
+            mut key_fragments: NibbleVec
+        ) -> Option<&'a $($mut_)* Trie<K, V> > where K: TrieKey {
+            // Handle retrieval at the root.
+            if key_fragments.len() == 0 {
+                println!("hello");
+                return Some(trie);
+            }
+
+            let bucket = key_fragments.get(0) as usize;
+
+            match trie.children[bucket] {
+                None => None,
+                Some(ref $($mut_)* existing_child) => {
+                    match match_keys(&key_fragments, &existing_child.key) {
+                        KeyMatch::Full => Some(existing_child),
+                        KeyMatch::Partial(idx) => {
+                            let new_key_fragments = key_fragments.split(idx);
+                            $name(& $($mut_)* *existing_child, new_key_fragments)
+                        },
+                        KeyMatch::FirstPrefix => Some(existing_child),
```

Comment:
Now that I think about the cases I'm tempted to make a method called `get_closest_descendant` with the same behaviour as your `get_node` method! Here's a full enumeration of the cases for looking up the closest descendants of a key `l`, compared to some stored key `s`.

```
l: abcde
s: abcde (KeyMatch::Full - closest descendants are children with the shortest keys).
s: abced (KeyMatch::Partial - no meaningful closest descendants).
s: abcdefgh (KeyMatch::FirstPrefix - this node is the *only* close descendant).
s: abc (KeyMatch::SecondPrefix - no meaningful closest descendants).
```

In the full match case we have a choice between returning the exactly matching node itself (a sort of "greater than or equal to" behaviour) or the children themselves...